include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
+    )
+    {
+        // TODO: can we optimize?
+        H = computeMicrofacetNormal<vector3_type>(transmitted,V,L,orientedEta);
+        retval.NdotH = nbl::hlsl::dot<vector3_type>(N, H);
+
+        // not coming from the medium (reflected) OR
+        // exiting at the macro scale AND ( (not L outside the cone of possible directions given IoR with constraint VdotH*LdotH<0.0) OR (microfacet not facing toward the macrosurface, i.e. non heightfield profile of microsurface) )
+        const bool valid = !transmitted || (VdotL <= -min(orientedEta, rcpOrientedEta) && retval.NdotH > nbl::hlsl::numeric_limits<scalar_type>::min);
+        if (valid)
+        {
+            // TODO: can we optimize?
+            retval.VdotH = nbl::hlsl::dot<vector3_type>(V,H);
+            retval.LdotH = nbl::hlsl::dot<vector3_type>(L,H);
+            retval.NdotH2 = retval.NdotH * retval.NdotH;
+            return true;
+        }
+        return false;
+    }
+    template<class ObserverRayDirInfo, class IncomingRayDirInfo NBL_FUNC_REQUIRES(ray_dir_info::Basic<ObserverRayDirInfo> && ray_dir_info::Basic<IncomingRayDirInfo>)
+    static bool compute(
+        NBL_REF_ARG(this_t) retval,
+        NBL_CONST_REF_ARG(surface_interactions::SIsotropic<ObserverRayDirInfo>) interaction,
+        NBL_CONST_REF_ARG(SLightSample<IncomingRayDirInfo>) _sample,
+        const scalar_type eta, NBL_REF_ARG(vector3_type) H
+    )
+    {
+        const scalar_type NdotV = interaction.getNdotV();
+        const scalar_type NdotL = _sample.getNdotL();
+        const bool transmitted = isTransmissionPath(NdotV,NdotL);
+
+        fresnel::OrientedEtas<scalar_type> orientedEta = fresnel::OrientedEtas<scalar_type>::create(NdotV, eta);
+
+        const vector3_type V = interaction.getV().getDirection();
+        const vector3_type L = _sample.getL().getDirection();
+        const scalar_type VdotL = nbl::hlsl::dot<vector3_type>(V, L);
+        return compute(retval,transmitted,V,L,interaction.getN(),NdotL,VdotL,orientedEta.value,orientedEta.rcp,H);
+    }
+    template<class ObserverRayDirInfo, class IncomingRayDirInfo NBL_FUNC_REQUIRES(ray_dir_info::Basic<ObserverRayDirInfo> && ray_dir_info::Basic<IncomingRayDirInfo>)
+    static bool compute(
+        NBL_REF_ARG(this_t) retval,
+        NBL_CONST_REF_ARG(surface_interactions::SIsotropic<ObserverRayDirInfo>) interaction,
+        NBL_CONST_REF_ARG(SLightSample<IncomingRayDirInfo>) _sample,
+        const scalar_type eta
+    )
+    {
+        vector3_type dummy;
+        return compute<ObserverRayDirInfo, IncomingRayDirInfo>(retval,interaction,_sample,eta,dummy);
+    }
+
+    bool isValidVNDFMicrofacet(const bool is_bsdf, const bool transmission, const scalar_type VdotL, const scalar_type eta, const scalar_type rcp_eta)
+    {
+        return NdotH >= 0.0 && !(is_bsdf && transmission && (VdotL > -min(eta, rcp_eta)));
+    }
+
+    scalar_type getVdotH() NBL_CONST_MEMBER_FUNC { return VdotH; }
```

Comment:
because of how `computeMicrofacetNormal` works you could put an `assert(VdotH>=scalar_type(0));` in here